symcache/src/writer.rs

```diff
@@ -563,20 +565,23 @@ where
         Ok(original)
     }
 
+    fn ensure_sorted(&mut self) {
+        // Only sort if functions do not already appear in order. They are sorted primarily by their
+        // start address, and secondarily by the index in which they appeared originally in the
+        // file.
+        if !self.sorted {
+            dmsort::sort_by_key(&mut self.functions, |handle| handle.original);
+        }
+    }
+
     fn write_functions(&mut self) -> Result<format::Seg<format::FuncRecord>, SymCacheError> {
         if self.functions.is_empty() {
             return Ok(format::Seg::default());
         }
 
-        // Only sort if functions do not already appear in order. They are sorted primarily by their
-        // start address, and secondarily by the index in which they appeared originally in the
-        // file.
         // To compute parent offsets after that, one can simply binary search by the parent_ref
         // handle, allowing for efficient unique lookups.
-        if !self.sorted {
-            self.functions
-                .sort_unstable_by_key(|handle| handle.original);
-        }
+        self.ensure_sorted();
```

Comment:
Worst case, we might sort twice here, even though it's not necessary. However, we cannot reset the `sorted` flag since we need to fix up inline functions, and it should be fast enough in general.